oak/server/rust/oak_runtime/src/node/http/mod.rs

```diff
@@ -0,0 +1,520 @@
+            .expect("Error when reading request body from the connection")
+            .to_vec();
+
+        HttpRequest {
+            url,
+            method,
+            body,
+            headers,
+        }
+    }
+}
+
+/// Oak Node implementation for the HTTP server.
+impl Node for HttpServerNode {
+    fn run(
+        self: Box<Self>,
+        runtime: RuntimeProxy,
+        startup_handle: oak_abi::Handle,
+        notify_receiver: oneshot::Receiver<()>,
+    ) {
+        // At start-of-day we need/expect to receive a write handle for an invocation channel
+        // to use for all subsequent activity.
+        info!("{}: Waiting for invocation channel", self.node_name);
+        let invocation_channel =
+            match HttpServerNode::try_get_invocation_channel(&runtime, startup_handle) {
+                Ok(writer) => writer,
+                Err(status) => {
+                    error!(
+                        "Failed to retrieve invocation channel write handle: {:?}",
+                        status
+                    );
+                    return;
+                }
+            };
+        if let Err(err) = runtime.channel_close(startup_handle) {
+            error!(
+                "Failed to close initial inbound channel {}: {:?}",
+                startup_handle, err
+            );
+        }
+
+        // Build a service to process all incoming HTTP requests.
+        let generic_handler = HttpRequestHandler {
+            runtime,
+            invocation_channel,
+        };
+        let server = self.make_server(generic_handler, notify_receiver);
+
+        // Create an Async runtime for executing futures.
+        // https://docs.rs/tokio/
+        let mut async_runtime = create_async_runtime();
+
+        // Start the HTTP server.
+        info!(
+            "{}: Starting HTTP server pseudo-Node on: {}",
+            self.node_name, self.address
+        );
+        async_runtime.block_on(server);
+    }
+}
+
+fn create_async_runtime() -> tokio::runtime::Runtime {
+    tokio::runtime::Builder::new()
+        // Use simple scheduler that runs all tasks on the current-thread.
+        // https://docs.rs/tokio/0.2.16/tokio/runtime/index.html#basic-scheduler
+        .basic_scheduler()
+        // Enables the I/O driver.
+        // Necessary for using net, process, signal, and I/O types on the Tokio runtime.
+        .enable_io()
+        // Enables the time driver.
+        // Necessary for creating a Tokio Runtime.
+        .enable_time()
+        .build()
+        .expect("Couldn't create Async runtime")
+}
+
+/// [`HttpRequestHandler`] handles HTTP requests from a client and sends HTTP responses back.
+#[derive(Clone)]
+struct HttpRequestHandler {
+    /// Reference to the Runtime in the context of this HTTP server pseudo-Node.
+    runtime: RuntimeProxy,
+    /// Channel handle used for writing HTTP invocations.
+    invocation_channel: oak_abi::Handle,
+}
+
+impl HttpRequestHandler {
+    fn handle(
```

Comment:
Why can this not be an async fn?

Reply:
What was the compiler error when using an async function and removing the `async move`?

Reply:
I see. I don't think there is an easy fix. I think it could be fixed by manually implementing the `MakeSvc` trait (https://docs.rs/hyper/0.13.7/hyper/service/index.html#makeservice rather) than using `make_service_fn` but I don't think it is worth the effort to just remove this one `async move`.

Reply:
I agree :)